Update kube-dns and nodelocaldns to 1.23.0

DamianSawicki opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

The PR bumps `kube-dns` and `nodelocaldns` to the latest version `1.23.0`. The new version of kube-dns includes dnsmasq v2.90, which (among other things)
 * fixes the long-standing memory leak for SRV records with zero TTL, and
 * introduces the flag --max-tcp-connections to adjust the previously hardcoded limit on the number of concurrent TCP connections.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
N/A

#### Special notes for your reviewer:

`kube-dns 1.23.0` upgrades `dnsmasq` to `v2.90`.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

DamianSawicki committed:
Update kube-dns and nodelocaldns to 1.23.0

Mostly to upgrade dnsmasq to v2.90.

DamianSawicki commented:
/cc @aojea

aojea commented:
/lgtm
/approve

@DamianSawicki please remove the release note, this changes are only used by some CI jobs, are not exposed to the users

DamianSawicki commented:
> /lgtm /approve
> 
> @DamianSawicki please remove the release note, this changes are only used by some CI jobs, are not exposed to the users

Thanks for reviewing! I removed the release note as instructed and moved the relevant info to _What this PR does / why we need it_.

aojea commented:
Thanks, this can wait for next release as is an internal CI details
/milestone v1.31

dims commented:
/approve
/lgtm
